cli/Makefile

```
@@ -12,6 +12,8 @@ LOADER_LDFLAGS = $(JLDFLAGS) -ffreestanding -L$(build_shlibdir) -L$(build_libdir
 
 ifeq ($(OS),WINNT)
 LOADER_CFLAGS += -municode -mconsole -nostdlib -fno-stack-check -fno-stack-protector -mno-stack-arg-probe
+else ifeq ($(OS),Linux)
+LOADER_CFLAGS += -DLIBSTDCXXPROBE
```

We typically use the existing `__linux__` define in the file, rather than adding this to the makefile. We can can try to add @ararslan (FreeBSD) to this too by making the detection more accurate also:
```suggestion
#ifndef LIBSTDCXXPROBE
#ifdef defined(__GLIBC__)
#define LIBSTDCXXPROBE 1
#elseif
#define LIBSTDCXXPROBE 0
#endif
#endif
```

---

(I think the empty `#elseif` there should just be an `#else`)

---

#34276 is typically a problem only on Linux: there you can expect to have a new libstdc++ which system libraries may have been linked to.  I don't think this is the case on FreeBSD?

---

I believe FreeBSD uses libcxx, libsdtc++ is only if a dep needs it.

---

Ah, true. But checking if you are using glibc should be right in any case for deciding this, I think?

---

On second thought, yes, I think checking for `__GLIBC__` should work just fine.  I was the one who asked Aaron to make this into a Makefile define, because I figured we might need to do something more complicated than the C preprocessor can handle, and I also thought we might want a Makefile switch to disable this functionality permanently, but I think checking for `__GLIBC__` and also having a runtime environment variable escape hatch is sufficient.

---

Ah, that's a great point @giordano, I got confused between glibc and libstdc++.  You're right, it's not enough to check `__GLIBC__`, we need to basically check if our compiler is GCC and we're on Linux instead.

The alternative, of course, is to simply do this everywhere, since we do ship our own `libstdc++` everywhere.  That's not the end of the world, we just have to gracefully handle not finding a global `libstdc++`.

---

> we need to basically check if our compiler is GCC and we're on Linux instead.

I'd say only Linux, which is what I initially did: https://github.com/JuliaLang/julia/blob/ef1f5dc6751d4c302add4525f56bbdb04063448c/cli/loader_lib.c#L173-L176 :innocent:

We can build with Clang on Linux, but still have libstdc++ in the search path.

---

I thought there might be Linux distros that use clang/libcxx by default, but it appears there aren't any other than OpenMandriva.  I found these slides which are kind of a fun read: https://llvm.org/devmtg/2019-04/slides/TechTalk-Rosenkranzer-Switching_a_Linux_distro.pdf

That being said, yes, `_OS_LINUX_` should just be enough here for now.  :)

---

Reverted the makefile change, back to just checking `_OS_LINUX_`.